File: src/core/sdk/typescript/src/services/ml-governance.ts
/**
 * ACGS-2 ML Governance Service
 * Adaptive ML models with feedback loops and drift detection
 * Constitutional Hash: cdd01ef066bc6cf2
 */

import type { ACGS2Client } from '../client/index.js';
import { CONSTITUTIONAL_HASH } from '../types/index.js';
import type {
  ABNTest,
  CreateABNTestRequest,
  CreateMLModelRequest,
  DriftDetection,
  FeedbackSubmission,
  MLModel,
  MakePredictionRequest,
  ModelPrediction,
  PaginatedResponse,
  SubmitFeedbackRequest,
  UpdateMLModelRequest,
} from '../types/index.js';

export class MLGovernanceService {
  private client: ACGS2Client;
  private basePath = '/api/v1/ml-governance';

  constructor(client: ACGS2Client) {
    this.client = client;
  }

  async createModel(request: CreateMLModelRequest): Promise<MLModel> {
    const response = await this.client.post(`${this.basePath}/models`, {
      ...request,
      constitutionalHash: CONSTITUTIONAL_HASH,
    });
    return response.data as MLModel;
  }

  async getModel(modelId: string): Promise<MLModel> {
    const response = await this.client.get(`${this.basePath}/models/${modelId}`);
    return response.data as MLModel;
  }

  async listModels(params?: {
    page?: number;
    pageSize?: number;
    modelType?: string;
    framework?: string;
  }): Promise<PaginatedResponse<MLModel>> {
    const queryParams = new URLSearchParams();
    if (params?.page) queryParams.append('page', params.page.toString());
    if (params?.pageSize) queryParams.append('pageSize', params.pageSize.toString());
    if (params?.modelType) queryParams.append('modelType', params.modelType);
    if (params?.framework) queryParams.append('framework', params.framework);

    const response = await this.client.get(`${this.basePath}/models?${queryParams.toString()}`);
    return response.data as PaginatedResponse<MLModel>;
  }

  async updateModel(modelId: string, request: UpdateMLModelRequest): Promise<MLModel> {
    const response = await this.client.put(`${this.basePath}/models/${modelId}`, {
      ...request,
      constitutionalHash: CONSTITUTIONAL_HASH,
    });
    return response.data as MLModel;
  }

  async deleteModel(modelId: string): Promise<void> {
    await this.client.delete(`${this.basePath}/models/${modelId}`);
  }

  async makePrediction(request: MakePredictionRequest): Promise<ModelPrediction> {
    const response = await this.client.post(`${this.basePath}/predictions`, {
      ...request,
      constitutionalHash: CONSTITUTIONAL_HASH,
    });
    return response.data as ModelPrediction;
  }

  async getPrediction(predictionId: string): Promise<ModelPrediction> {
    const response = await this.client.get(`${this.basePath}/predictions/${predictionId}`);
    return response.data as ModelPrediction;
  }

  async listPredictions(params?: {
    modelId?: string;
    page?: number;
    pageSize?: number;
  }): Promise<PaginatedResponse<ModelPrediction>> {
    const queryParams = new URLSearchParams();
    if (params?.modelId) queryParams.append('modelId', params.modelId);
    if (params?.page) queryParams.append('page', params.page.toString());
    if (params?.pageSize) queryParams.append('pageSize', params.pageSize.toString());

    const response = await this.client.get(`${this.basePath}/predictions?${queryParams.toString()}`);
    return response.data as PaginatedResponse<ModelPrediction>;
  }

  async submitFeedback(request: SubmitFeedbackRequest): Promise<FeedbackSubmission> {
    const response = await this.client.post(`${this.basePath}/feedback`, {
      ...request,
      constitutionalHash: CONSTITUTIONAL_HASH,
    });
    return response.data as FeedbackSubmission;
  }

  async getFeedback(feedbackId: string): Promise<FeedbackSubmission> {
    const response = await this.client.get(`${this.basePath}/feedback/${feedbackId}`);
    return response.data as FeedbackSubmission;
  }

  async listFeedback(params?: {
    modelId?: string;
    feedbackType?: string;
    page?: number;
    pageSize?: number;
  }): Promise<PaginatedResponse<FeedbackSubmission>> {
    const queryParams = new URLSearchParams();
    if (params?.modelId) queryParams.append('modelId', params.modelId);
    if (params?.feedbackType) queryParams.append('feedbackType', params.feedbackType);
    if (params?.page) queryParams.append('page', params.page.toString());
    if (params?.pageSize) queryParams.append('pageSize', params.pageSize.toString());

    const response = await this.client.get(`${this.basePath}/feedback?${queryParams.toString()}`);
    return response.data as PaginatedResponse<FeedbackSubmission>;
  }

  async checkDrift(modelId: string): Promise<DriftDetection> {
    const response = await this.client.get(`${this.basePath}/models/${modelId}/drift`);
    return response.data as DriftDetection;
  }

  async retrainModel(
    modelId: string,
    options?: { feedbackThreshold?: number }
  ): Promise<Record<string, unknown>> {
    const payload: Record<string, unknown> = {
      constitutionalHash: CONSTITUTIONAL_HASH,
    };
    if (options?.feedbackThreshold) {
      payload['feedbackThreshold'] = options.feedbackThreshold;
    }

    const response = await this.client.post(`${this.basePath}/models/${modelId}/retrain`, payload);
    return response.data as Record<string, unknown>;
  }

  async createABNTest(request: CreateABNTestRequest): Promise<ABNTest> {
    const response = await this.client.post(`${this.basePath}/ab-tests`, {
      ...request,
      constitutionalHash: CONSTITUTIONAL_HASH,
    });
    return response.data as ABNTest;
  }

  async getABNTest(testId: string): Promise<ABNTest> {
    const response = await this.client.get(`${this.basePath}/ab-tests/${testId}`);
    return response.data as ABNTest;
  }

  async listABNTests(params?: {
    page?: number;
    pageSize?: number;
  }): Promise<PaginatedResponse<ABNTest>> {
    const queryParams = new URLSearchParams();
    if (params?.page) queryParams.append('page', params.page.toString());
    if (params?.pageSize) queryParams.append('pageSize', params.pageSize.toString());

    const response = await this.client.get(`${this.basePath}/ab-tests?${queryParams.toString()}`);
    return response.data as PaginatedResponse<ABNTest>;
  }

  async stopABNTest(testId: string): Promise<ABNTest> {
    const response = await this.client.post(`${this.basePath}/ab-tests/${testId}/stop`);
    return response.data as ABNTest;
  }

  async getABNTestResults(testId: string): Promise<Record<string, unknown>> {
    const response = await this.client.get(`${this.basePath}/ab-tests/${testId}/results`);
    return response.data as Record<string, unknown>;
  }

  async getModelMetrics(
    modelId: string,
    params?: { startDate?: string; endDate?: string }
  ): Promise<Record<string, unknown>> {
    const queryParams = new URLSearchParams();
    if (params?.startDate) queryParams.append('startDate', params.startDate);
    if (params?.endDate) queryParams.append('endDate', params.endDate);

    const response = await this.client.get(
      `${this.basePath}/models/${modelId}/metrics?${queryParams.toString()}`
    );
    return response.data as Record<string, unknown>;
  }

  async getDashboardData(): Promise<Record<string, unknown>> {
    const response = await this.client.get(`${this.basePath}/dashboard`);
    return response.data as Record<string, unknown>;
  }
}

File: src/core/sdk/python/acgs2_sdk/services/ml_governance.py
"""
ACGS-2 ML Governance Service
Adaptive ML models with feedback loops and drift detection
Constitutional Hash: cdd01ef066bc6cf2
"""

from typing import TYPE_CHECKING, Any

try:
    from src.core.shared.types import JSONDict, JSONValue
except ImportError:
    JSONValue = str | int | float | bool | None | dict[str, Any] | list[Any]  # type: ignore[misc]
    JSONDict = dict[str, JSONValue]  # type: ignore[misc]

from acgs2_sdk.constants import CONSTITUTIONAL_HASH, ML_GOVERNANCE_ENDPOINT
from acgs2_sdk.models import (
    ABNTest,
    CreateABNTestRequest,
    CreateMLModelRequest,
    DriftDetection,
    FeedbackSubmission,
    MakePredictionRequest,
    MLModel,
    ModelPrediction,
    PaginatedResponse,
    SubmitFeedbackRequest,
    UpdateMLModelRequest,
)

if TYPE_CHECKING:
    from acgs2_sdk.client import ACGS2Client


class MLGovernanceService:
    """Service for ML model governance and adaptive learning."""

    def __init__(self, client: "ACGS2Client") -> None:
        self._client = client
        self._base_path = ML_GOVERNANCE_ENDPOINT

    async def create_model(
        self,
        request: CreateMLModelRequest,
    ) -> MLModel:
        """Create/register a new ML model."""
        data = await self._client.post(
            f"{self._base_path}/models",
            json={
                **request.model_dump(by_alias=True, exclude_none=True),
                "constitutionalHash": CONSTITUTIONAL_HASH,
            },
        )
        return MLModel.model_validate(data.get("data", data))

    async def get_model(self, model_id: str) -> MLModel:
        """Get an ML model by ID."""
        data = await self._client.get(f"{self._base_path}/models/{model_id}")
        return MLModel.model_validate(data.get("data", data))

    async def list_models(
        self,
        page: int = 1,
        page_size: int = 50,
        model_type: str | None = None,
        framework: str | None = None,
    ) -> PaginatedResponse[MLModel]:
        """List ML models."""
        params: JSONDict = {"page": page, "pageSize": page_size}
        if model_type:
            params["modelType"] = model_type
        if framework:
            params["framework"] = framework

        data = await self._client.get(f"{self._base_path}/models", params=params)
        response_data = data.get("data", data)
        return PaginatedResponse[MLModel](
            data=[MLModel.model_validate(m) for m in response_data.get("data", [])],
            total=response_data.get("total", 0),
            page=response_data.get("page", page),
            page_size=response_data.get("pageSize", page_size),
            total_pages=response_data.get("totalPages", 0),
        )

    async def update_model(
        self,
        model_id: str,
        request: UpdateMLModelRequest,
    ) -> MLModel:
        """Update an ML model."""
        data = await self._client.put(
            f"{self._base_path}/models/{model_id}",
            json={
                **request.model_dump(by_alias=True, exclude_none=True),
                "constitutionalHash": CONSTITUTIONAL_HASH,
            },
        )
        return MLModel.model_validate(data.get("data", data))

    async def delete_model(self, model_id: str) -> None:
        """Delete an ML model."""
        await self._client.delete(f"{self._base_path}/models/{model_id}")

    async def make_prediction(
        self,
        request: MakePredictionRequest,
    ) -> ModelPrediction:
        """Make a prediction with an ML model."""
        data = await self._client.post(
            f"{self._base_path}/predictions",
            json={
                **request.model_dump(by_alias=True, exclude_none=True),
                "constitutionalHash": CONSTITUTIONAL_HASH,
            },
        )
        return ModelPrediction.model_validate(data.get("data", data))

    async def get_prediction(self, prediction_id: str) -> ModelPrediction:
        """Get a prediction by ID."""
        data = await self._client.get(f"{self._base_path}/predictions/{prediction_id}")
        return ModelPrediction.model_validate(data.get("data", data))

    async def list_predictions(
        self,
        model_id: str | None = None,
        page: int = 1,
        page_size: int = 50,
    ) -> PaginatedResponse[ModelPrediction]:
        """List model predictions."""
        params: JSONDict = {"page": page, "pageSize": page_size}
        if model_id:
            params["modelId"] = model_id

        data = await self._client.get(f"{self._base_path}/predictions", params=params)
        response_data = data.get("data", data)
        return PaginatedResponse[ModelPrediction](
            data=[ModelPrediction.model_validate(p) for p in response_data.get("data", [])],
            total=response_data.get("total", 0),
            page=response_data.get("page", page),
            page_size=response_data.get("pageSize", page_size),
            total_pages=response_data.get("totalPages", 0),
        )

    async def submit_feedback(
        self,
        request: SubmitFeedbackRequest,
    ) -> FeedbackSubmission:
        """Submit feedback for model training."""
        data = await self._client.post(
            f"{self._base_path}/feedback",
            json={
                **request.model_dump(by_alias=True, exclude_none=True),
                "constitutionalHash": CONSTITUTIONAL_HASH,
            },
        )
        return FeedbackSubmission.model_validate(data.get("data", data))

    async def get_feedback(
        self,
        feedback_id: str,
    ) -> FeedbackSubmission:
        """Get feedback by ID."""
        data = await self._client.get(f"{self._base_path}/feedback/{feedback_id}")
        return FeedbackSubmission.model_validate(data.get("data", data))

    async def list_feedback(
        self,
        model_id: str | None = None,
        feedback_type: str | None = None,
        page: int = 1,
        page_size: int = 50,
    ) -> PaginatedResponse[FeedbackSubmission]:
        """List feedback submissions."""
        params: JSONDict = {"page": page, "pageSize": page_size}
        if model_id:
            params["modelId"] = model_id
        if feedback_type:
            params["feedbackType"] = feedback_type

        data = await self._client.get(f"{self._base_path}/feedback", params=params)
        response_data = data.get("data", data)
        return PaginatedResponse[FeedbackSubmission](
            data=[FeedbackSubmission.model_validate(f) for f in response_data.get("data", [])],
            total=response_data.get("total", 0),
            page=response_data.get("page", page),
            page_size=response_data.get("pageSize", page_size),
            total_pages=response_data.get("totalPages", 0),
        )

    async def check_drift(self, model_id: str) -> DriftDetection:
        """Check for model drift."""
        data = await self._client.get(f"{self._base_path}/models/{model_id}/drift")
        return DriftDetection.model_validate(data.get("data", data))

    async def retrain_model(
        self,
        model_id: str,
        feedback_threshold: int | None = None,
    ) -> JSONDict:
        """Trigger model retraining."""
        json_data: JSONDict = {"constitutionalHash": CONSTITUTIONAL_HASH}
        if feedback_threshold:
            json_data["feedbackThreshold"] = feedback_threshold

        data = await self._client.post(
            f"{self._base_path}/models/{model_id}/retrain",
            json=json_data,
        )
        return data.get("data", data)

    async def create_ab_test(
        self,
        request: CreateABNTestRequest,
    ) -> ABNTest:
        """Create an A/B test."""
        data = await self._client.post(
            f"{self._base_path}/ab-tests",
            json={
                **request.model_dump(by_alias=True, exclude_none=True),
                "constitutionalHash": CONSTITUTIONAL_HASH,
            },
        )
        return ABNTest.model_validate(data.get("data", data))

    async def get_ab_test(self, test_id: str) -> ABNTest:
        """Get an A/B test by ID."""
        data = await self._client.get(f"{self._base_path}/ab-tests/{test_id}")
        return ABNTest.model_validate(data.get("data", data))

    async def list_ab_tests(
        self,
        page: int = 1,
        page_size: int = 50,
    ) -> PaginatedResponse[ABNTest]:
        """List A/B tests."""
        params = {"page": page, "pageSize": page_size}
        data = await self._client.get(f"{self._base_path}/ab-tests", params=params)
        response_data = data.get("data", data)
        return PaginatedResponse[ABNTest](
            data=[ABNTest.model_validate(t) for t in response_data.get("data", [])],
            total=response_data.get("total", 0),
            page=response_data.get("page", page),
            page_size=response_data.get("pageSize", page_size),
            total_pages=response_data.get("totalPages", 0),
        )

    async def stop_ab_test(self, test_id: str) -> ABNTest:
        """Stop an A/B test."""
        data = await self._client.post(f"{self._base_path}/ab-tests/{test_id}/stop")
        return ABNTest.model_validate(data.get("data", data))

    async def get_ab_test_results(self, test_id: str) -> JSONDict:
        """Get A/B test results."""
        data = await self._client.get(f"{self._base_path}/ab-tests/{test_id}/results")
        return data.get("data", data)

    async def get_model_metrics(
        self,
        model_id: str,
        start_date: str | None = None,
        end_date: str | None = None,
    ) -> JSONDict:
        """Get model performance metrics."""
        params: JSONDict = {}
        if start_date:
            params["startDate"] = start_date
        if end_date:
            params["endDate"] = end_date

        data = await self._client.get(f"{self._base_path}/models/{model_id}/metrics", params=params)
        return data.get("data", data)

    async def get_dashboard_data(self) -> JSONDict:
        """Get ML governance dashboard data."""
        data = await self._client.get(f"{self._base_path}/dashboard")
        return data.get("data", data)
